---
title: "CNS Tumour Classification Report"
output:
  pdf_document:
    latex_engine: pdflatex
  html_document:
    df_print: paged
header-includes:
- \usepackage{palatino}
- \renewcommand{\familydefault}{\sfdefault} % sans serif
- \fontfamily{ppl}\selectfont
geometry: margin=0.2in
always_allow_html: true
---
<style type="text/css">
  body{
  font-size: 14pt;
  text-align: justify;
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr)
library(kableExtra)
library(ggplot2)
```

```{r check_platform, echo=FALSE}
platform <- "Unknown"
if (seq!="false"){
    platform <- seq
}
if (seq=="F") {
    platform <- "MinION/GridION"
}
if (seq=="P") {
    platform <- "PromethION"
}
if (seq=="P2S") {
    platform <- "P2 Solo"
}
if (seq=="GXB") {
    platform <- "GridION"
}
```

```{r rapidcns2_setup, echo=FALSE}
rapidcns2_run <- FALSE
### if rapidcns2 has been run
if (!rf_details=="NULL" & !votes=="NULL")
rapidcns2_run <- TRUE
```

```{r sturgeon_setup, echo=FALSE}
sturgeon_run <- FALSE
### if sturgeon has been run
if (!sturgeon_csv=="NULL")
sturgeon_run <- TRUE
```

```{r nanodx_setup, echo=FALSE}
nanodx_run <- FALSE
### if nanodx has been run
if (!nanodx_votes=="NULL")
nanodx_run <- TRUE
```

```{r analysis_string, echo=FALSE}
classifier_run <- FALSE
analysis_string <- list()
if ( rapidcns2_run == "TRUE" ) {
    analysis_string <- c(analysis_string, " the rapidCNS^2^ pipeline (Patel *et al.*, 2022),")
    classifier_run <- TRUE
}
if ( sturgeon_run == "TRUE" ) {
    analysis_string <- c(analysis_string, " the Sturgeon pipeline (Vermeulen *et al.*, 2023),")
    classifier_run <- TRUE
}
if ( nanodx_run == "TRUE" ) {
    classifier_run <- TRUE
    analysis_string <- c(analysis_string, " the nanoDx/crossNN pipeline (Yuan *et al.*, 2024; Kuschel *et al.*, 2023; Euskirchen *et al.*, 2017),")
}

n_methods <- length(analysis_string)
count = 0
output_string <- ""
for (row in analysis_string) {
  count = count + 1
  if (count != n_methods) {
    output_string <- paste0(output_string, row)
  } else {
    if (n_methods != 1) {
    output_string <- paste0(output_string, " and", row)
    } else {
      output_string <- row
    }
  }
}
analysis_string <- paste0(substr(output_string, 1, nchar(output_string)-1),".")

```

```{r, echo=FALSE, warning=FALSE, results='asis', eval=rapidcns2_run}
x <- rf_details
oob <- read.delim(x,header = F, sep = ":")
mc <- read.delim(votes,sep=" ")
sites <- as.integer(oob[1,2])
brierscore <- round(oob[3,2]*100,digits=1)
rapid_cns2_top_class <- rownames(mc[which.max(mc$cal_Freq),])
rapid_cns2_top_conf <- round(mc$cal_Freq[which.max(mc$cal_Freq)],digits=1)
```

```{r, echo=FALSE, warning=FALSE, results='asis', eval=sturgeon_run}
sturgeon_scores <- read.table(sturgeon_csv, sep=",")
sturgeon_scores <- as.data.frame(t(sturgeon_scores))
sturgeon_scores <- sturgeon_scores[-1,]
colnames(sturgeon_scores) <- c("class", "score")
sturgeon_scores$score <- as.numeric(sturgeon_scores$score)
sturgeon_top_conf <- round(sturgeon_scores[order(-sturgeon_scores$score),][1,]$score, 2)
sturgeon_top_class <- sturgeon_scores[order(-sturgeon_scores$score),][1,]$class
```

```{r, echo=FALSE, warning=FALSE, results='asis', eval=nanodx_run}
nanodx_scores <- read.table(nanodx_votes, skip=1, sep="\t")
colnames(nanodx_scores) <- c("V1", "class", "score", "num_features")
nanodx_scores$score <- as.numeric(nanodx_scores$score)
nanodx_top_conf <- round(nanodx_scores[order(-nanodx_scores$score),][1,]$score, 2)
nanodx_top_class <- nanodx_scores[order(-nanodx_scores$score),][1,]$class
```

```{r,echo=FALSE, ,warning=FALSE}
df <- read.delim(coverage)
df$Reads <- "All"
df$Reads[endsWith(df$chrom,"_region")] <- "On_target"
global_cov <- round(mean(df[which(df$Reads == "All"),]$mean), 2)
on_target_cov <- round(mean(df[which(df$Reads == "On_target"),]$mean), 2)
```

## **Sample ID:** `r sample` \

#### This report was generated by the **methylation based CNS tumour classification pipeline (NEXTFLOW CNS_tumour_characterisation `r nextflow_ver`)** and is intended for research use only.
#### Tumour tissue was prepared and processed using a modified ULK114 protocol (Oxford Nanopore Technologies). 160 gene regions from the NPHD panel were targeted using an adaptive sampling approach on the `r platform` platform. \
#### Analysis and/or classification of CNS tumours by methylation data using this tool should be strictly considered **for research only** and the method is under active development. It has not been clinically validated in sufficiently large cohorts. Interpretation and implementation of the results in a clinical setting is in the sole responsibility of the treating physician.

```{r, echo=FALSE, results='asis', eval=classifier_run}
cat("#### Data was analysed by ",  analysis_string)
```

```{r, echo=FALSE, results='asis', eval=classifier_run}
cat("## <strong> Sample Classification Summary</strong>")
```

```{r, echo=FALSE, results='asis', eval=classifier_run}
cat("#### This sample has the following classifications (and confidence scores):")
```

```{r, echo=FALSE, results='asis', eval=rapidcns2_run}
cat("#### Rapid_CNS^2^ classification: <strong>", rapid_cns2_top_class, " (", rapid_cns2_top_conf, ").</strong>", sep="")
```

```{r, echo=FALSE, results='asis', eval=sturgeon_run}
cat("#### Sturgeon classification: <strong>", sturgeon_top_class, " (", sturgeon_top_conf, ")</strong>.", sep="")
```

```{r, echo=FALSE, results='asis', eval=nanodx_run}
cat("#### NanoDx classification: <strong>", nanodx_top_class, " (", nanodx_top_conf, ")</strong>.", sep="")
```

## **Sample Coverage Summary**
```{r, echo=FALSE, results='asis'}
cat("#### Global Estimated Average Depth of Coverage: <strong>", global_cov, "x</strong>.", sep="")
```

```{r, echo=FALSE, results='asis'}
cat("#### Targets Estimated Average Depth of Coverage: <strong>", on_target_cov, "x</strong>.", sep="")
```


## **On-target coverage**
```{r,echo=FALSE,out.width="75%",warning=FALSE, fig.align='center'}
df <- read.delim(coverage)
df$Reads <- "All"
df$Reads[endsWith(df$chrom,"_region")] <- "On_target"
df$chrom <- gsub("_region","",df$chrom)
df$mean <- as.numeric(df$mean)
df$chrom[df$chrom=="chr1"] <- "chr01"
df$chrom[df$chrom=="chr2"] <- "chr02"
df$chrom[df$chrom=="chr3"] <- "chr03"
df$chrom[df$chrom=="chr4"] <- "chr04"
df$chrom[df$chrom=="chr5"] <- "chr05"
df$chrom[df$chrom=="chr6"] <- "chr06"
df$chrom[df$chrom=="chr7"] <- "chr07"
df$chrom[df$chrom=="chr8"] <- "chr08"
df$chrom[df$chrom=="chr9"] <- "chr09"

df <- df[order(df$chrom),]

df$chrom <- as.character(df$chrom)

df <- df[which(df$chrom!="chrM"),]

df$chrom <- as.factor(df$chrom)
levels(df$chrom) <- unique(df$chrom)
ggplot2::ggplot(data=df, 
  aes(x=chrom,y=mean,color=Reads)) + 
  geom_point(size=3) + 
  theme_minimal() + 
  ylab("Mean coverage") + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1,size = 10), axis.title.y = element_text(size=10)) + 
  ylim(0,max(df$mean))+ scale_color_manual(values = c("#FF5A5F","#0B3954")) + 
  xlab("Chromosome") +theme(text=element_text(family="sans")) + 
  theme(panel.border = element_rect(color = "black", 
                                    fill = NA, 
                                    size = 1))
```
\
\newpage

---

## **Mutations**

````{=html}
```{r, echo=FALSE, results='asis', eval=inc_igvreport}
xfun::file_string(igv_report)
```
````


```{r,echo=FALSE, out.width="100%", warning=FALSE, eval=exc_igvreport, results='asis'}
d <- read.csv(mutations)
kable(d, booktabs = TRUE, caption="ALL MUTATIONS") %>%
  kable_styling(font_size=14, latex_options = c("HOLD_position","scale_down"))
```

\
\newpage

```{r,echo=FALSE, out.width="100%", warning=FALSE, eval=exc_igvreport, results='asis'}
if ( file.exists(somatic_mutations) ) {

  count_lines <- function(file_path) {
  lines <- readLines(file_path, warn = FALSE)
  return(length(lines))
  }

  # Count the number of lines in the file
  num_lines <- count_lines(somatic_mutations)

  if ( num_lines > 0 ) {
  d <- read.csv(somatic_mutations)
  kable(d, booktabs = TRUE, caption="SOMATIC MUTATIONS") %>%
  kable_styling(font_size=14, latex_options = c("HOLD_position","scale_down"))
  }
} else {
cat("#### No somatic variants were found in this sample.")
}
```

\
\newpage

---

## **Copy number profile**

```{r, echo=FALSE, out.width="100%",warning=FALSE}
knitr::include_graphics(cnv_plot)
```
\newpage
\
	
---

```{r, echo=FALSE, warning=FALSE, results='asis', eval=rapidcns2_run}
cat("## <strong> Rapid CNS^2^ Methylation classification </strong>")
```


```{r, echo=FALSE, warning=FALSE, results='asis', eval=rapidcns2_run}
cat("#### The rapidCNS^2^ (random forest) classification is \"", rapid_cns2_top_class, "\" with a confidence score of ", rapid_cns2_top_conf, ". The score distribution for the Top 10 entities is given below.", sep="")
```

```{r,echo=FALSE,warning=FALSE, results='asis', eval=rapidcns2_run}
mc$methClass <- rownames(mc)
ggplot(data=mc, aes(x=methClass, y=cal_Freq)) +
  geom_bar(stat="identity", fill="#dc143c") +
  coord_flip() +
  scale_x_discrete(limits=mc[order(mc$cal_Freq,decreasing=T),]$methClass[10:1]) +
  ylim(0,100) +
  ylab("confidence score (%)") + xlab("methylation class") +
  theme_classic() + theme(aspect.ratio=1) +theme(text=element_text(family="sans")) +
  theme(panel.border = element_rect(color = "black", fill = NA, size = 1))

```

\
\newpage

```{r, echo=FALSE, warning=FALSE, results='asis', eval=sturgeon_run}
cat("## <strong> Sturgeon Methylation Classification </strong>")
```

```{r, echo=FALSE, warning=FALSE, results='asis', eval=sturgeon_run}
cat("#### The Sturgeon classification is \"", sturgeon_top_class, "\" with a confidence score of ", sturgeon_top_conf, ". The score distribution for the Top 10 entities is given below.", sep="")
```

```{r,echo=FALSE,warning=FALSE, eval=sturgeon_run}
ggplot(data=sturgeon_scores, aes(x=class, y=score)) +
  geom_bar(stat="identity", fill="#dc143c") +
  coord_flip() +
  scale_x_discrete(limits=sturgeon_scores[order(sturgeon_scores$score,decreasing=T),]$class[10:1]) +
  ylim(0,1) +
  ylab("confidence score (%)") + xlab("methylation class") +
  theme_classic() + theme(aspect.ratio=1) +theme(text=element_text(family="sans")) +
  theme(panel.border = element_rect(color = "black", fill = NA, size = 1))
```


```{r, echo=FALSE, warning=FALSE, results='asis', eval=nanodx_run}
cat("## <strong>nanoDx Methylation Classification </strong>")
```

```{r, echo=FALSE, warning=FALSE, results='asis', eval=nanodx_run}
cat("#### The NanoDx classification is \"", nanodx_top_class, "\" with a confidence score of ", nanodx_top_conf, ". The score distribution for the Top 10 entities is given below.", sep="")
```

```{r,echo=FALSE,warning=FALSE, eval=nanodx_run}
ggplot(data=nanodx_scores, aes(x=class, y=score)) +
  geom_bar(stat="identity", fill="#dc143c") +
  coord_flip() +
  scale_x_discrete(limits=nanodx_scores[order(nanodx_scores$score,decreasing=T),]$class[10:1]) +
  ylim(0,1) +
  ylab("confidence score (%)") + xlab("methylation class") +
  theme_classic() + theme(aspect.ratio=1) +theme(text=element_text(family="sans")) +
  theme(panel.border = element_rect(color = "black", fill = NA, size = 1))
```

\newpage
\

---

## **MGMT Promoter Methylation** \
#### Average coverage of MGMT is: **`r cov`**x.\

```{r setup2, echo=FALSE}
show_mgmt <- FALSE
mgmt_too_low <- TRUE
### if methylartist and mgmt exist
if (methylartist_plot == "true" & mgmt == "true")
if (as.numeric(mgmt_cov) >= as.numeric(mgmt_minimum_cov))
show_mgmt <- TRUE
mgmt_too_low <- FALSE
```


```{r conditional_block, echo=FALSE, results='asis', eval=show_mgmt}
mgmt_file <- read.csv(opt$mgmt)
avg <- mgmt_file$average
avg <- round(avg, 2)
avg <- paste0(avg,"%")
status <- mgmt_file$status
cat("#### MGMT promoter methylation status was assessed using the 'rapidCNS2' method. The average methylation cutoff, above which a sample is classified as methylated, is 25%.")
cat("\n\n#### Average methylation = ")
cat(avg)
cat("\n\n#### Predicted MGMT promoter status = ")
cat(status)
```

```{r, echo=FALSE, warning=FALSE, results='asis', eval=mgmt_too_low}
cat("#### MGMT promoter methylation status was not assessed in this sample due to sequence coverage at the MGMT promter region being too low.")
```

## **MGMT Promoter Methylation Plot:**
```{r, echo=FALSE, out.width="100%",warning=FALSE, eval=show_mgmt}
knitr::include_graphics(opt$methylartist)
```

```{r, echo=FALSE, warning=FALSE, results='asis', eval=mgmt_too_low}
cat("#### An MGMT promoter plot was not produced in this sample due to sequence coverage at the MGMT site being too low.")
```

```{r, echo=FALSE, warning=FALSE, results='asis'}
avg_cov <- round(as.numeric(read.csv(avg_gene_cov)), 2)
```

---

## **Genes Coverage**
#### The following table identifies potential outliers differing significantly (>1.5 standard deviations) from the mean coverage of `r avg_cov`x.
```{r, echo=FALSE, warning=FALSE, results='asis'} 
if ( file.exists(uniq_genes_cov) ) {
  d <- read.csv(uniq_genes_cov)
  kable(d, booktabs = TRUE) %>%
  kable_styling(font_size=14, latex_options = c("HOLD_position","scale_down")) %>%
  row_spec(0, background = NULL, bold = FALSE, color = NULL)
} else {
cat("#### Error producing genes coverage table.")
}
```
---

## **Software Versions**
```{r, echo=FALSE, warning=FALSE, results='asis'} 
if ( file.exists(software_versions) ) {
  d <- read.csv(software_versions)
  kable(d, booktabs = TRUE) %>%
  kable_styling(font_size=14, latex_options = c("HOLD_position","scale_down")) %>%
  row_spec(0, background = NULL, bold = FALSE, color = NULL)
} else {
cat("#### Error producing software versions table.")
}
```
---

## **User Specified Parameters**
```{r, echo=FALSE, warning=FALSE, results='asis'} 
if ( file.exists(user_params) ) {
  d <- read.csv(user_params)
  kable(d, booktabs = TRUE) %>%
  kable_styling(font_size=14, latex_options = c("HOLD_position","scale_down")) %>%
  row_spec(0, background = NULL, bold = FALSE, color = NULL)
} else {
cat("#### Error producing user parameters table.")
}
```

---

## **Important Information** \

#### The tumour classification and report generation is based on, and builds upon, the rapidCNS^2^ protocol designed and published by Patel _et_ _al._ (2022).

#### Patel, A., Dogan, H., Payne, A., Krause, E., Sievers, P., Schoebe, N., Schrimpf, D., Blume, C., Stichel, D., Holmes, N., Euskirchen, P., Hench, J., Frank, S., Rosenstiel-Goidts, V., Ratliff, M., Etminan, N., Unterberg, A., Dieterich, C., Herold-Mende, C., Pfister, S. M., … Sahm, F. (2022). Rapid-CNS^2^: rapid comprehensive adaptive nanopore-sequencing of CNS tumors, a proof-of-concept study. _Acta_ _neuropathologica_, _143_(5), 609–612. https://doi.org/10.1007/s00401-022-02415-6 \

#### The Sturgeon classification was performed using the Sturgeon classifier (https://github.com/marcpaga/sturgeon) designed and published by Vermeulen _et_ _al._ (2023). \ 

#### C. Vermeulen, M. Pagès-Gallego, L. Kester, M. E. G. Kranendonk, P. Wesseling, N. Verburg, P. de Witt Hamer, E. J. Kooi, L. Dankmeijer, J. van der Lugt, K. van Baarsen, E. W. Hoving, B. B. J. Tops & J. de Ridder (2023) Ultra-fast deep-learned CNS tumour classification during surgery. _Nature,_ 622, 842-849. https://doi.org/10.1038/s41586-023-06615-2  \ 

#### The nanoDx/crossNN classification was performed using the nanoDx classifier (https://gitlab.com/pesk/nanoDx) designed and published by Vermeulen _et_ _al._ (2023). \ 

#### Yuan, D., Jugas, R., Pokorna, P., Sterba, J., Slaby, O., Schmid, S., Siewert, C., Osberg, B., Capper, D., Zeiner, P., Weber, K., Harter, P., Jabareen, N., Mackowiak, S., Ishaque, N., Eils, R., Lukassen, S., & Euskirchen, P. (2024). An explainable framework for cross-platform DNA methylation-based classification of cancer (S. 2024.01.22.24301523). medRxiv. (pre-print)

#### Kuschel LP, Hench J, Frank S, Hench IB, Girard E, Blanluet M, Masliah-Planchon J, Misch M, Onken J, Czabanka M, Yuan D, Lukassen S, Karau P, Ishaque N, Hain EG, Heppner F, Idbaih A, Behr N, Harms C, Capper D, Euskirchen P. Robust methylation-based classification of brain tumours using nanopore sequencing. _Neuropathol Appl Neurobiol._ 2023 Feb;49(1):e12856. doi: 10.1111/nan.12856.

#### Euskirchen P, Bielle F, Labreche K, Kloosterman WP, Rosenberg S, Daniau M, Schmitt C, Masliah-Planchon J, Bourdeaut F, Dehais C, Marie Y, Delattre JY, Idbaih A. Same-day genomic and epigenomic diagnosis of brain tumors using real-time nanopore sequencing. _Acta Neuropathol._ 2017 Nov;134(5):691-703. doi: 10.1007/s00401-017-1743-5.

#### **The information contained with this report is intended for research use only and is not for use in diagnostic procedures.**

#### Date: `r format(Sys.time(), '%d %B, %Y')`

